### ISSUE TITLE
Problem using dates as factors

### ISSUE THREAD

**Issue by imbhe** [rvlenth/emmeans#162]
I got erroneous results using emmeans and pairs on models using dates as factors. An example can be found at https://chalmersuniversity.box.com/s/bvfrngnu0fyg3pblbtabqlyw9o8l9wp7.

**Comment by rvlenth** [comment]
I am paraphrasing the bug in a reproducible way so that others looking at this can tell what it is. 

The dataset:
```r
> df
# A tibble: 12 x 5
   Date       DateStr  Block Trt    Conc
   <date>     <chr>    <int> <fct> <dbl>
 1 2014-01-15 20140115     1 PS     3.58
 2 2014-01-15 20140115     1 NM     4.84
 3 2014-01-15 20140115     1 C      5.37
 4 2014-01-15 20140115     2 C      3.42
 5 2014-01-15 20140115     2 PS     3.81
 6 2014-01-15 20140115     2 NM     3.3 
 7 2014-02-17 20140217     1 PS     3.12
 8 2014-02-17 20140217     1 NM     3.66
 9 2014-02-17 20140217     1 C      5.25
10 2014-02-17 20140217     2 C      2.71
11 2014-02-17 20140217     2 PS     2.93
12 2014-02-17 20140217     2 NM     2.84
```
Models considered: (Note: Please don't re-use the same object name for two different models in a bug report)
```r
mod1 <- lm(log(Conc) ~ Trt + as.factor(Block) + as.factor(Date), data = df)
mod2 <- lm(log(Conc) ~ Trt + as.factor(Block) + as.factor(DateStr), data = df)
```
These two models have the same predictions:
```r
> library(emmeans)
> predict(ref_grid(mod1)) - predict(ref_grid(mod2))
 [1] 0 0 0 0 0 0 0 0 0 0 0 0
```
However, the EMMs differ:
```r
> emmeans(mod1, "Trt")
 Trt emmean     SE df lower.CL upper.CL
 C     1.33 0.0566  7     1.20     1.47
 NM    1.30 0.0566  7     1.17     1.43
 PS    1.24 0.0566  7     1.11     1.38

Results are averaged over the levels of: Block, Date 
Results are given on the log (not the response) scale. 
Confidence level used: 0.95 

> emmeans(mod2, "Trt")
 Trt emmean   SE df lower.CL upper.CL
 C     1.39 0.08  7     1.20     1.58
 NM    1.28 0.08  7     1.09     1.47
 PS    1.21 0.08  7     1.02     1.40

Results are averaged over the levels of: Block, DateStr 
Results are given on the log (not the response) scale. 
Confidence level used: 0.95 
```

This is indeed a strange result. I will look into it.

**Comment by rvlenth** [comment]
This was a subtle one to track down, but I figured out what happened. There is code to make sure the grid is in a standard order before computing EMMs. That code involves ordering by the levels of factors and non-factors. In the case of `mod1`, the levels of the non-factor `Date` are of class `Date` rather than `character`, and that caused a glitch whereby the ordering variable became a vector of  `NA` values, producing essentially a random ordering. I have fixed it now:
```r
> emmeans(mod1, "Trt")
 Trt emmean   SE df lower.CL upper.CL
 C     1.39 0.08  7     1.20     1.58
 NM    1.28 0.08  7     1.09     1.47
 PS    1.21 0.08  7     1.02     1.40

Results are averaged over the levels of: Block, Date 
Results are given on the log (not the response) scale. 
Confidence level used: 0.95 
```
I will push the repaired version to github when I can, but am stymied by some change in **roxygen2** that totally messes up my `Rd` files. When I get that straightened out, I'll push this and reference this issue.

Thanks for reporting it.
